lib/fhir_models/validation/element_validator/data_type_validator.rb

```diff
@@ -0,0 +1,79 @@
+    module DataTypeValidator
+      # FHIR Resources, Profiles and the StructureDefinitions are made up of FHIR Data Types.
+      # There are two kinds of structures that fall under the FHIR Data Types: complex-type and primitive-type.
+      # The snapshot of a resource does not contain the element definitions associated with primitive-type or complex-type
+      # structures it is composed of.
+      #
+      # https://www.hl7.org/fhir/datatypes.html
+      # @param element [Object] The Element of the Resource under test
+      # @param element_definition [FHIR::ElementDefinition] The Element Definition from which the cardinality is taken
+      # @return result [FHIR::ValidationResult] The result of the cardinality check
+      def self.validate(resource, element_definition)
+        return unless element_definition.path.include? '.' # Root Elements do not have a type
+
+        return if element_definition.type.empty?
+
+        elements = FHIR::Validation::Retrieval.retrieve_by_element_definition(resource,
+                                                                              element_definition,
+                                                                              indexed: true)
+
+        elements.flat_map do |path, el|
+          validate_element(el, element_definition, path)
+        end
+      end
+
+      def self.validate_element(element, element_definition, path)
+        # Get the type
+        type_code = if element_definition.type.one?
+                      element_definition.type.first.code
+                    else
+                      return UnknownType.new('Need path in order to determine type') unless path
+
+                      element_definition.type.find do |datatype|
+                        cap_code = "#{datatype.code[0].capitalize}#{datatype.code[1..-1]}"
+                        /[^.]+$/.match(element_definition.path.gsub('[x]', cap_code)).to_s == /[^.]+$/.match(path).to_s
+                      end.code
+                    end
+        type_def = FHIR::Definitions.type_definition(type_code) || FHIR::Definitions.resource_definition(type_code)
+
+        # If we are missing the Structure Definition needed to do the validation.
+        if type_def.nil?
+          return FHIR::ValidationResult.new(element_definition: element_definition,
+                                            validation_type: :datatype,
+                                            result: :warn,
+                                            text: (if type_code.nil?
+                                                     'Type code not provided or extends the primitive code'
+                                                   else
+                                                     "Unknown type: #{type_code}"
+                                                   end),
+                                            element_path: path || element_definition.path)
+        elsif type_def.kind == 'primitive-type'
```

Comment:
It would happen often...on all the primitives.  I removed it and just return on these since this information isn't that useful (unlike a warning that we didn't check a code because we were missing a validator).